### PR TITLE
Show distributed performance tests in Teamcity

### PR DESCRIPTION
**Pull request by wolfs**
This change has the effect to show the distributed
performance tests as regular tests in Teamcity - with
duration and exception message.

Here is an example what it looks like: https://builds.gradle.org/viewLog.html?buildId=1956221&buildTypeId=Gradle_Branches_Performance_PerformanceTestCoordinatorLinux&tab=testsInfo

And here is a failing performance test: https://builds.gradle.org/viewLog.html?buildId=1952279&tab=buildResultsDiv&buildTypeId=Gradle_Branches_Performance_PerformanceTestCoordinatorLinux

### CODE REVIEW

File: buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        testListener.afterSuite(testSuiteDescriptor.parent, new DefaultTestResult(TestResult.ResultType.SUCCESS, 0, 0, 0, 0, 0, []))
+        testListener.afterSuite(testSuiteDescriptor.parent.parent, new DefaultTestResult(TestResult.ResultType.SUCCESS, 0, 0, 0, 0, 0, []))
+    }
+
+    private void publishAdditionalMetadata(DecoratingTestDescriptor testCaseDescriptor, Object build) {
+        def buildUrl = build.@webUrl
+        if (buildUrl) {
+            testOutputListener.onOutput(testCaseDescriptor, new DefaultTestOutputEvent(TestOutputEvent.Destination.StdOut,
+                """
+                    Scenario: ${getScenarioId(build)}
```
Comment (oehme) [inline buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy]: If we can also get the link to the performance report in here, then we can completely get rid of the "scenario report" code.

File: buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    private String getScenarioId(Object build) {
+        NodeChildren properties = build.properties.children()
+        properties.find { it.@name == 'scenario' }.@value.text()
+    }
+
+
+    void addTestListener(TestListener listener) {
+        if (!listener.getClass().getName().startsWith('org.gradle.api.internal')) {
```
Comment (oehme) [inline buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy]: I guess this is to suppress the events from the "actual" test execution?

Comment (wolfs) [inline buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy]: It is not suppressing the events (that is done by not calling `super.addTestListener`) but ignoring the listeners added by the test task itself - these would be responsible for generating an html and xml report. Even worse, one of the `TestOutputListener`s (`TestReportDataCollector`) would try to write to the `outputWriter` which is closed after the "actual" test execution is finished.

Comment (oehme) [inline buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy]: Got it. Could you add this summary to the class documentation?

File: buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy (hunk 1 above)
Comment (wolfs) [inline buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy]: I added the link. See https://builds.gradle.org/viewLog.html?buildId=1964638&tab=buildResultsDiv&buildTypeId=Gradle_Branches_Performance_PerformanceTestCoordinatorLinux

**Comment by wolfs** [comment]
@oehme I addressed your comments. Could you have another look?

**Comment by oehme** [comment]
Thank you Stefan, this is such a cool improvement :)

**Pull request merged**